bouncer/views.py

```diff
@@ -25,10 +30,27 @@ def annotation(self):
                 id=self.request.matchdict["id"])
         except exceptions.NotFoundError:
             statsd.incr("views.annotation.404.annotation_not_found")
-            raise httpexceptions.HTTPNotFound()
+            raise httpexceptions.HTTPNotFound("Annotation not found")
+
+        try:
+            annotation_id = document["_id"]
+            document_uri = document["_source"]["uri"]
+        except KeyError:
+            raise httpexceptions.HTTPUnprocessableEntity(
+                _("The annotation is invalid"))
+
+        if not isinstance(document_uri, str):
+            raise httpexceptions.HTTPUnprocessableEntity(
+                _("The annotation has an invalid document URI"))
```

Comment:
I've moved this comment over from #10, since it's more appropriate here.

After spending some quality time sifting through all the special ways we store annotations at the moment, I found quite a few that don't have a `uri`, and even some that don't have anything that could be used as a URI.
The Postgres migration code for example first tries to read the URI from `target[0][source]`, if that doesn't exist it tries to fall back to `uri`, and if taht one doesn't exist it refuses to migrate the annotation.

My suggestion for bouncer is something like:

1. Read from `target[0][source]`
2. When still missing or not a http(s) uri, try `uri`
3. Maybe: When still missing, find the first http(s) uri from `target[0].link` with `rel == 'canonical'` (careful, sometimes `target[0].link` is not a list but a string)
4. When all of the above fail, render an error